packages/parser/src/parser.ts

```diff
@@ -20,52 +24,88 @@ interface ParseForESLintResult {
   scopeManager: ReturnType<typeof analyzeScope>;
 }
 
+function validateBoolean(
+  value: boolean | undefined,
+  fallback: boolean = false
+): boolean {
+  if (typeof value !== 'boolean') {
+    return fallback;
+  }
+  return value;
+}
+
+//------------------------------------------------------------------------------
+// Public
+//------------------------------------------------------------------------------
+
 export const version = packageJSON.version;
 
+export const Syntax = Object.freeze(AST_NODE_TYPES);
+
 export function parse(code: string, options?: ParserOptions) {
   return parseForESLint(code, options).ast;
 }
 
-export const Syntax = Object.freeze(typescriptESTree.AST_NODE_TYPES);
-
-export function parseForESLint<T extends ParserOptions = ParserOptions>(
+export function parseForESLint(
   code: string,
-  options?: T | null
+  options?: ParserOptions | null
 ): ParseForESLintResult {
-  if (typeof options !== 'object' || options === null) {
-    options = { useJSXTextNode: true } as T;
-  } else if (typeof options.useJSXTextNode !== 'boolean') {
-    options = Object.assign({}, options, { useJSXTextNode: true });
-  }
-  if (typeof options.filePath === 'string') {
-    const tsx = options.filePath.endsWith('.tsx');
-    if (tsx || options.filePath.endsWith('.ts')) {
-      options = Object.assign({}, options, { jsx: tsx });
-    }
+  if (!options || typeof options !== 'object') {
+    options = {};
   }
-
   // https://eslint.org/docs/user-guide/configuring#specifying-parser-options
   // if sourceType is not provided by default eslint expect that it will be set to "script"
-  options.sourceType = options.sourceType || 'script';
   if (options.sourceType !== 'module' && options.sourceType !== 'script') {
     options.sourceType = 'script';
   }
+  if (typeof options.ecmaFeatures !== 'object') {
+    options.ecmaFeatures = {};
+  }
+
+  const parserOptions: ParserOptionsTsESTree = {
+    useJSXTextNode: validateBoolean(options.useJSXTextNode, true),
+    range: validateBoolean(options.range, true), // TODO: this option is ignored by typescript-estree
+    loc: validateBoolean(options.loc, true), // TODO: this option is ignored by typescript-estree
+    tokens: validateBoolean(options.tokens, true),
+    jsx: validateBoolean(options.ecmaFeatures.jsx),
+    comment: validateBoolean(options.comment, true),
+    errorOnUnknownASTType: validateBoolean(options.errorOnUnknownASTType),
+    errorOnTypeScriptSyntacticAndSemanticIssues: validateBoolean(
+      options.errorOnTypeScriptSyntacticAndSemanticIssues
+    )
+  };
+
+  if (options.project) {
+    parserOptions.project = options.project;
+  }
+
+  if (typeof options.tsconfigRootDir === 'string') {
+    parserOptions.tsconfigRootDir = options.tsconfigRootDir;
+  }
+
+  if (options.extraFileExtensions) {
+    parserOptions.extraFileExtensions = options.extraFileExtensions;
```

Comment:
we should extract this from `ext` of eslint if users uses custom file extensions